specify lifecycle of stateful fault tolerance strategies

Ladicek opened the pull request:
Signed-off-by: Ladislav Thon <lthon@redhat.com>

Resolves #479.

Ladicek commented:
This is a draft, because I need to [figure out how to] add some TCK tests. However, feedback about the specification wording is welcome!

Emily-Jiang commented on spec/src/main/asciidoc/bulkhead.asciidoc:
```diff
@@ -71,8 +71,19 @@ public Future<Connection> serviceA() {
+==== Lifecycle
+
+Unlike some other fault tolerance strategies, bulkhead needs to store some data between invocations: the number of currently running executions, or the queue of waiting executions.
+This state is a singleton, irrespective of the lifecycle of the bean that uses the `@Bulkhead` annotation.
```
It is associated with the class instead of the instances.

Ladicek replied:
I've changed it to use both explanations.

Emily-Jiang commented on spec/src/main/asciidoc/bulkhead.asciidoc:
```diff
@@ -71,8 +71,19 @@ public Future<Connection> serviceA() {
+
+Unlike some other fault tolerance strategies, bulkhead needs to store some data between invocations: the number of currently running executions, or the queue of waiting executions.
+This state is a singleton, irrespective of the lifecycle of the bean that uses the `@Bulkhead` annotation.
+If, for example, the bean is `@RequestScoped` and therefore there is a new instance of the bean for each request, all instances share the same bulkead state.
```
blukead -> bulkhead

Emily-Jiang commented on spec/src/main/asciidoc/bulkhead.asciidoc:
```diff
@@ -71,8 +71,19 @@ public Future<Connection> serviceA() {
 
+==== Lifecycle
+
+Unlike some other fault tolerance strategies, bulkhead needs to store some data between invocations: the number of currently running executions, or the queue of waiting executions.
```
Delete "Unlike some other fault tolerance strategies," as Circuit Breaker has the same stategy.

Emily-Jiang commented on spec/src/main/asciidoc/circuitbreaker.asciidoc:
```diff
@@ -109,6 +109,15 @@ public void service() {
 
+==== Lifecycle
+
+Unlike some other fault tolerance strategies, circuit breaker needs to store some data between invocations: the number of recent successful and failed invocations, or how long has the circuit breaker been open.
```
Delete "Unlike some other fault tolerance strategies," as Bulkhead has the same stategy.

Ladicek commented:
I tried to reword it slightly, based on your comments. Thanks @Emily-Jiang!

Azquelt commented on spec/src/main/asciidoc/bulkhead.asciidoc:
```diff
@@ -71,8 +71,20 @@ public Future<Connection> serviceA() {
+For example, if the bean is `@RequestScoped` and therefore there is a new instance of the bean for each request, all instances share the same bulkhead state.
+
+More specifically: each method of each class that uses a bulkhead has its own state.
+If a subclass inherits a `@Bulkhead` method, then the subclass has its own, distinct, bulkhead state.
```
I found this part surprising.

If a method is inherited by a subclass, why should it receive a distinct bulkhead state for the subclass? I has assumed that each method declaration would have a separate state, rather than each method declaration and subclass combination having a separate state.

Do you see advantages to defining it one way or the other?

I would see having one state per method declaration as consistent with the way we use annotations to apply fault tolerance policies to methods. When deciding which policies to apply, you must find the declaration of the method which is to be run and then consider:
 * annotations applied to the method
 * annotations applied to the class which declares the method
 * annotations applied to superclasses of the class which declares the method (because our annotations have `@Inherited`)

You ignore:
 * annotations applied to any method that the method overrides
 * annotations applied to any subclass of the class which declares the method

Ladicek replied:
I think it was you who pointed out that subclasses should have their own circuit breakers (applies to bulkhead as well): https://github.com/eclipse/microprofile-fault-tolerance/issues/479#issuecomment-561197781

But you raise very good points. I'll try to address them (tomorrow).

Azquelt replied:
Heh, when I was writing this, it did ring a few bells.

Unfortunately, I suspect whichever way round we do it, we might catch someone out :(

I think on balance, I'm leaning towards the state being per method declaration for the reasons set out above. It's also fairly easy to override the method and delegate to `super` if you do want a separate bulkhead state.

Ladicek replied:
I didn't forget about this, I just had a ton of other work. I got back to it today, and changed to your suggestion. I agree that tying the state to the method declaration makes most sense.

I'm wondering if something happens to a `@CircuitBreaker` method if a subclass which _doesn't_ override it adds a `@CircuitBreaker` annotation. I'd say no, but I'd also say that this should be specified somewhere for all annotations. Same thing with configuration overrides -- can I override configuration of a `@CircuitBreaker` method on a subclass, even if that subclass doesn't override the method? This is something that should be specified for all annotations. (I'm not sure if it is or not, didn't check :-) )

Ladicek commented:
I have started adding TCK tests for this. For now, I started with circuit breaker, and will continue with bulkhead later, when we agree on desired behavior for CB.

The current TCK tests mirror existing tests for usage of `@Retry` in inheritance hierarchy, both in their structure and their expectations. The specification wording hence is _not_ aligned to what the current TCK tests expect.

I present this for discussion, not as a definitive version that is ready for merge.

Azquelt commented:
> The current TCK tests mirror existing tests for usage of `@Retry` in inheritance hierarchy, both in their structure and their expectations. The specification wording hence is _not_ aligned to what the current TCK tests expect.

Bah, yes, it looks like back in #183 it was decided that annotations on the current class should be considered, rather than just the ones on the declaring class.

I didn't realise that. Obviously we can't have circuitbreakers with different settings sharing the same state so either we'd have to change the existing behaviour, or redefine how the state is shared here.

I don't really like the idea of changing the existing behaviour since I think it's fairly likely that it'll break existing code.

Ladicek commented:
Yea. I need to re-read the spec to figure out the right wording for "where" is the CB state located.

Emily-Jiang commented:
We discussed today we agreed the circuit breaker scope is per class per method.

Ladicek commented:
I have just adjusted the specification wording, it should be now in sync with the TCK tests.

Now I need to add more TCK tests.

Ladicek commented:
I have added a few more tests for circuit breaker, and added some tests for bulkhead. I don't think it's necessary to test all the inheritance hierarchy scenarios with bulkhead, because they have been tested with retry for a long time, and are now tested with circuit breaker as well.

Emily-Jiang commented on spec/src/main/asciidoc/bulkhead.asciidoc:
```diff
@@ -71,8 +71,21 @@ public Future<Connection> serviceA() {
 
+==== Lifecycle
+
+Bulkhead needs to needs to maintain some state between invocations: the number of currently running executions, or the queue of waiting executions.
```
dup of `needs to`

Ladicek replied:
Thanks!

Ladicek replied:
Done.

Emily-Jiang commented on spec/src/main/asciidoc/bulkhead.asciidoc:
```diff
@@ -71,8 +71,21 @@ public Future<Connection> serviceA() {
+For example, if the bean is `@RequestScoped` and therefore there is a new instance of the bean for each request, all instances share the same bulkhead state.
+
+More specifically: the bulkhead state is bound to the bean class _and_ the guarded method.
+If all the bulkheads were stored in a `Map`, then the key to the `Map` would be a combination of the bean class (`java.lang.Class`) and the method object (`java.lang.reflect.Method`).
```
I would reword to be:
The combination of the bean class (`java.lang.Class`) and the method object (`java.lang.reflect.Method`) uniquely identify a bulkhead.

Ladicek replied:
Done.

Emily-Jiang commented on spec/src/main/asciidoc/bulkhead.asciidoc:
```diff
@@ -71,8 +71,21 @@ public Future<Connection> serviceA() {
+
+More specifically: the bulkhead state is bound to the bean class _and_ the guarded method.
+If all the bulkheads were stored in a `Map`, then the key to the `Map` would be a combination of the bean class (`java.lang.Class`) and the method object (`java.lang.reflect.Method`).
+Implementations are of course free to use any strategy they wish.
```
This can be deleted.

Ladicek replied:
Done.

Emily-Jiang commented on spec/src/main/asciidoc/circuitbreaker.asciidoc:
```diff
@@ -110,6 +110,17 @@ public void service() {
+Circuit breaker needs to maintain some state between invocations: the number of recent successful and failed invocations, or how long has the circuit breaker been open.
+This state is a singleton, irrespective of the lifecycle of the bean that uses the `@CircuitBreaker` annotation.
+In other words, the state is associated with the bean class, not with its instances.
+For example, if the bean is `@RequestScoped` and therefore there is a new instance of the bean for each request, all instances share the same circuit breaker state.
```
This line suggests one circuit breaker per class. The following paragraph states it is associated with the class and the methods.

Ladicek replied:
Reworded.

Emily-Jiang commented on spec/src/main/asciidoc/circuitbreaker.asciidoc:
```diff
@@ -110,6 +110,17 @@ public void service() {
+For example, if the bean is `@RequestScoped` and therefore there is a new instance of the bean for each request, all instances share the same circuit breaker state.
+
+More specifically: the circuit breaker state is bound to the bean class _and_ the guarded method.
+If all the circuit breakers were stored in a `Map`, then the key to the `Map` would be a combination of the bean class (`java.lang.Class`) and the method object (`java.lang.reflect.Method`).
```
https://github.com/eclipse/microprofile-fault-tolerance/pull/525/files#r430455078

Emily-Jiang replied:
see above comments

Emily-Jiang commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/bulkhead/lifecycle/MutlipleMethodsBulkheadLifecycleService.java:
```diff
@@ -0,0 +1,37 @@
+        barrier.await();
+    }
+
+    @Bulkhead(8)
```
It works better if you use a different number.

Ladicek replied:
The purpose is to verify that they are different (even though they perhaps have the same number of allowed executions).

Ladicek commented:
I did some rewording based on @Emily-Jiang's suggestions (thanks!).

Azquelt commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/circuitbreaker/lifecycle/BaseCircuitBreakerOnMethod.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.concurrent.atomic.AtomicInteger;
+
+@Dependent
+@CBLifecycle(CBLifecycleServiceType.BASE_M)
```
I really don't like using a qualifier like this to distinguish between the test beans, it makes it much harder to navigate through the test code to find out which test is using which bean.

I realise that we already have some tests which do it this way. I assume this is because otherwise when one class extends another you get multiple beans which have the superclass as a bean type. If so, could we use `@Typed` to restrict the bean types instead.

I'd need to check it still works but I have a commit which replaces the qualifiers with `@Typed` instead.

Azquelt replied:
Here's the change I'd propose: https://github.com/Azquelt/microprofile-fault-tolerance/commit/828c2926be4ee51de216544ac242d1ad4f644eee

It does seem to work (in that all the tests fail in the expected way when run against open-liberty since we haven't implemented this yet)

Ladicek replied:
This is interesting, I didn't know about `@Typed`. I'll check.

Ladicek replied:
And yea, I copied what we do for the `@Retry` tests, even though I didn't like it. Your commit looks like a very nice simplification!

Ladicek replied:
Played with it a little and then just used your commit -- thanks!

Ladicek commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/bulkhead/lifecycle/AsyncInvoker.java:
```diff
@@ -0,0 +1,64 @@
+import static org.testng.Assert.fail;
+
+@ApplicationScoped
+public class AsyncInvoker {
```
I didn't know about `Exceptions.expect()`, it is indeed exactly the same and I'll use that.

When it comes to `AsyncCaller`, its `run` method (unlike `submit`) doesn't catch exceptins, and I'm not sure if that's intentional or mistake. So I created my own.

Ladicek replied:
Removed the assert method, but left `AsyncInvoker` for now.

Azquelt replied:
I had to look at this for a while before working out what I think is going on here.

I think `AsyncCaller.run()` doesn't catch exceptions because `@Asynchronous` will catch all thrown exceptions anyway.

`AsyncCaller.submit()` does catch exceptions because `Callable` throws checked exceptions so they have to be either caught or declared as thrown, even though `@Asynchronous` will always catch them.

Ladicek replied:
Ah, yea! Got it -- sorry it takes so long :-)

Ladicek commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/bulkhead/lifecycle/Barrier.java:
```diff
@@ -0,0 +1,43 @@
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+public class Barrier {
```
Do you know if TestNG has something akin to JUnit 4 rules? That's what I'd do if this was JUnit 4.

Azquelt replied:
So would I. I did look but TestNG doesn't seem to have anything equivalent. You could probably put together something with an `IInvokedMethodListener` which used reflection to find a field in the class but you'd have to check how it interacts with Arquillian.

The Arquillian extension mechanisms would be another option. I think you'd need to make a remote extension which observes before and after events, and then a local extension to make sure your remote extension gets packaged in each application. Again, you might need to check how that interacts with TestNG's lifecycle.

Lastly, we'd need some way of making what's going on clear to whoever is reading the tests.

A further thought: it might make sense to have an arquillian extension which fires CDI events at the start and end of each test method running on the server. Then we could implement any test fixtures as beans which observe those events.

I'm still on the fence a bit as to whether it's worth doing any of this rather than just having an `@AfterMethod` which adds a little more boilerplate in the test class, which is what we've done elsewhere.

Ladicek replied:
I'm thinking you not only need to open the barrier, you also need to wait for all the async operations to finish. That's why I did all the `try`/`catch`/`finally` dance. I'll try to think about this some more.

Azquelt replied:
Yeah, that may be true. I'll think on it some more too. For this PR, it's fine as-is.

Ladicek commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/bulkhead/lifecycle/BulkheadLifecycleTest.java:
```diff
@@ -0,0 +1,151 @@
+    private Instance<MutlipleMethodsBulkheadLifecycleService> multipleMethodsService;
+
+    @Test
+    public void noSharingBetweenClasses() throws InterruptedException, ExecutionException, TimeoutException {
```
Agree, will add the test.

Ladicek replied:
Added.

Ladicek commented:
When #548 is merged, I will add release notes to this PR as well.

Azquelt commented on tck/src/main/java/org/eclipse/microprofile/fault/tolerance/tck/bulkhead/lifecycle/MutlipleMethodsBulkheadLifecycleService.java:
```diff
@@ -0,0 +1,38 @@
+import javax.enterprise.context.Dependent;
+
+@Dependent
+public class MutlipleMethodsBulkheadLifecycleService {
+    @Bulkhead(8)
+    public void service1(Barrier barrier) {
+        barrier.await();
+    }
+
+    @Bulkhead(8)
+    public void service2(Barrier barrier) {
+        barrier.await();
+    }
```
Would this test be improved if the `@Bulkhead` annotation was on the class rather than the method?

The behaviour should be the same (separate bulkhead for each method).

Same question for `MutlipleMethodsCircuitBreakerLifecycleService`

Azquelt replied:
(Sorry, I missed this when I looked through it the first time).

Ladicek replied:
Hmm, yea, good point. I'll take a look.

Ladicek replied:
Done, thanks!

Ladicek commented:
Added release notes.
